### ISSUE TITLE
StepFailure is missing #respond_to?

### ISSUE THREAD

**Issue by flash-gordon** [dry-rb/dry-transaction#19]
Here is a snippet:

```ruby
transaction.fmap { happy_path }.or do |error|  
  errors = Array(error) # This fails with undefined method `to_ary' for "error message":String
end
```

Not to mention that `Array("error_message")` does not throw an exception by itself.

After spending a few minutes I figured out that `error` quacks like a String object but it is actually a StepFailure object. 
Adding `respond_to?` like

```ruby
      def respond_to?(*args)
        @__object.respond_to?(*args)
      end
```

fixes the problem but I do not _why_ (don't have time for digging into it atm) so I didn't sent a PR :)

**Comment by flash-gordon** [comment]
OK, I spent a dozen of minutes and that's what I figured out:
First of all I took a look at rubinius sources because you know ... C is not so welcome. rbx defines `Kernel#Array` [here](https://github.com/rubinius/rubinius/blob/20938ee86dcb7396c4cf2d16bb89ffd9fcee9a9d/core/kernel.rb#L99). So `Array` calls `Rubinius::Type.check_convert_type` which is described [here](https://github.com/rubinius/rubinius/blob/20938ee86dcb7396c4cf2d16bb89ffd9fcee9a9d/spec/core/type/check_convert_type_spec.rb#L4). Apparently it returns an object if type of the object defines a method either calls the method blindly.

After that I opened MRI sources and found a similar behavior. It tries to convert an object to array type [here](https://github.com/ruby/ruby/blob/931581469fcaa967c09ac1aaa5235f2ab5f65bd4/array.c#L660) and goes down to [this `respond_to?` check](https://github.com/ruby/ruby/blob/931581469fcaa967c09ac1aaa5235f2ab5f65bd4/vm_eval.c#L447) which fails so ruby calls `method_missing` [here](https://github.com/ruby/ruby/blob/931581469fcaa967c09ac1aaa5235f2ab5f65bd4/vm_eval.c#L460) and the later finally raises an exception. 

TL;DR you always should add `respond_to_missing?` to a `BasicObject` descendant if it is not a DSL object. Ruby itself does it [here](https://github.com/ruby/ruby/blob/trunk/lib/delegate.rb#L95).

I'll send a PR tomorrow.
